core/pta/ocall.c

```diff
@@ -0,0 +1,344 @@
+// SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2020, Microsoft Corporation
+ */
+
+#include <kernel/pseudo_ta.h>
+#include <kernel/thread.h>
+#include <mm/mobj.h>
+#include <optee_rpc_cmd.h>
+#include <pta_ocall.h>
+#include <tee/uuid.h>
+
+#define PTA_NAME "ocall.ta"
+
+#define PTR_ADD(ptr, offs) ((void *)((uintptr_t)(ptr) + (uintptr_t)(offs)))
+
+static TEE_Result compute_required_mobj_size(uint32_t param_types,
+					     TEE_Param params[TEE_NUM_PARAMS],
+					     size_t *required_size)
+{
+	TEE_Param *param;
+	uint32_t param_type;
+	size_t n;
+	size_t size = 0;
+
+	for (n = 0 ; n < TEE_NUM_PARAMS ; n++) {
+		param = params + n;
+		param_type = TEE_PARAM_TYPE_GET(param_types, n);
+
+		switch (param_type) {
+		case TEE_PARAM_TYPE_NONE:
+		case TEE_PARAM_TYPE_VALUE_INPUT:
+		case TEE_PARAM_TYPE_VALUE_INOUT:
+		case TEE_PARAM_TYPE_VALUE_OUTPUT:
+			break;
+		case TEE_PARAM_TYPE_MEMREF_INPUT:
+		case TEE_PARAM_TYPE_MEMREF_INOUT:
+		case TEE_PARAM_TYPE_MEMREF_OUTPUT:
+			if ((param->memref.buffer && !param->memref.size) ||
+			    (!param->memref.buffer && param->memref.size))
+				return TEE_ERROR_BAD_PARAMETERS;
+			if (ADD_OVERFLOW(size, param->memref.size, &size))
+				return TEE_ERROR_SECURITY;
+			break;
+		default:
+			return TEE_ERROR_BAD_PARAMETERS;
+		}
+	}
+
+	*required_size = size;
+	return TEE_SUCCESS;
+}
+
+static TEE_Result pre_process_params(struct thread_param *rpc_params,
+				     TEE_Param *ca_params,
+				     uint32_t ca_param_types,
+				     struct mobj *mobj, void *mobj_va)
+{
+	size_t n;
+	size_t mobj_offs = 0;
+
+	for (n = 0; n < TEE_NUM_PARAMS; n++) {
+		TEE_Param *ca_param = ca_params + n;
+		uint32_t ca_pt = TEE_PARAM_TYPE_GET(ca_param_types, n);
+
+		switch (ca_pt) {
+		case TEE_PARAM_TYPE_NONE:
+			rpc_params[n].attr = THREAD_PARAM_ATTR_NONE;
+			break;
+		case TEE_PARAM_TYPE_VALUE_INPUT:
```

Comment:
Ah, I like it; I wanted to get rid of the duplication but couldn't really due to the macro. I had seen code like what you suggested elsewhere in OP-TEE, but I figured that since the macro to set thread parameters was there, one is meant to use them.

Thank you for the suggestion.

Reply:
Added as suggested.